core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Planner rule that matches a
+ * {@link org.apache.calcite.rel.core.Join}
+ * and expands OR clauses in join conditions.
+ *
+ * <p>For example, for the following SQL query:
+ * <pre>
+ * SELECT *
+ * FROM t1
+ * JOIN t2
+ * ON (t1.id = t2.id OR t1.name = t2.name)
+ * </pre>
+ *
+ * <p>To
+ *
+ * <pre>
+ * SELECT *
+ * FROM t1
+ * JOIN t2
+ * ON t1.id = t2.id
+ * UNION ALL
+ * SELECT *
+ * FROM t1
+ * JOIN t2
+ * ON t1.name = t2.name
+ * </pre>
+ *
+ * <p>this rule would expand the OR condition into
+ * two separate join conditions, allowing the optimizer
+ * to handle these conditions more effectively.
+ */
+@Value.Enclosing
+public class JoinConditionOrExpansionRule
+    extends RelRule<JoinConditionOrExpansionRule.Config>
+    implements TransformationRule {
+
+  /** Creates an JoinConditionExpansionOrRule. */
+  protected JoinConditionOrExpansionRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Join join = call.rel(0);
+    RelBuilder relBuilder = call.builder();
+    List<RexNode> orConds = RelOptUtil.disjunctions(join.getCondition());
+
+    if (orConds.size() <= 1) {
+      return;
+    }
+
+    int leftFieldCount = join.getLeft().getRowType().getFieldCount();
+    for (RexNode cond : orConds) {
+      if (!isSimpleCondition(cond, leftFieldCount)) {
+        return;
+      }
+    }
+
+    List<RexNode> condsAppend = new ArrayList<>();
+    for (int i = 0; i < orConds.size(); i++) {
+      RexNode orCond = orConds.get(i);
+      for (int j = 0; j < i; j++) {
+        orCond = relBuilder.and(orCond, relBuilder.not(condsAppend.get(j)));
+      }
+      condsAppend.add(orCond);
+      relBuilder.push(join.getLeft())
+          .push(join.getRight())
+          .join(join.getJoinType(), orCond);
+    }
+
+    relBuilder.union(true, orConds.size());
+
+    final List<RexNode> projects = join.getRowType().getFieldList().stream()
```

Comment:
I didn't expect this to be necessary.
Why would types be different?

Reply:
I'll add a detailed error report here later.

Reply:
Strangely, I didn't get any error messages, and it worked fine without this logic. So I just removed it.